README
Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.


FirstBase
---------

FirstBase is an open source RDBMS and Application Builder.
It consists of a large set of end user programs designed to create,
manipulate, scan, organize, view, and format relational database and
database systems under the Unix/Linux environment.

The FirstBase tools are used to produce and maintain information tracking
applications and systems of data based on relational databases while still
maintaining the flexibility of Unix/Linux software tools.

Using a Unix/Linux tools approach creates an easily learned, dynamic,
and highly productive database management platform.

FirstBase features a unique approach to building database
applications. Using full screen dictionary and data editors, database
structures are defined and records of information are modified.

When desired, custom data views and entry forms can be created, set of data
can be organized using any combinations of fields, and tabular or free form
printouts can be generated.


Installation Instructions
-------------------------

See the file INSTALL for generalized installation instructions.
FirstBase uses the GNU `configure' method of generating system
dependent Makefiles in each source directory.

Note that the default installation is to install /usr/local/firstbase
as the FIRSTBASEHOME directory, with binaries in $FIRSTBASEHOME/bin,
libraries in $FIRSTBASEHOME/lib and scripts in $FIRSTBASEHOME/scripts.

Basically, this is something like

   ./configure
   make
   make install

If the `configure' script fails to generate workable Makefiles
on your system, please try and work out the options needed
for your particular system so I can fix them.

After installation, one environment variable needs to be added, 
and two PATH areas need to be modified.

FIRSTBASEHOME
-------------

The environment variable FIRSTBASEHOME needs to be added.

If you are using a c-shell style shell, use something like the following:

   setenv FIRSTBASEHOME /usr/local/firstbase

If you are using a Bourne type of shell, use something like this:

   FIRSTBASEHOME=/usr/local/firstbase
   export FIRSTBASEHOME

PATH
----

The PATH of each user needs to be modified to locate the firstbase
binaries and scripts. Using the FIRSTBASEHOME environment variable
is the best way to do this.

If you are using a c-shell style shell, use something like the following:

   set path=( $path $FIRSTBASEHOME/bin $FIRSTBASEHOME/scripts )

If you are using a Bourne type of shell, use something like this:

   PATH=$PATH:$FIRSTBASEHOME/bin:$FIRSTBASEHOME/scripts
   export PATH


MANPATH
-------

One last path to tweak if you want the man pages to be
available using the man command:

Add $FIRSTBASEHOME/man to the MANPATH environment variable.


Demonstration Applications
--------------------------

At this point, the `fb' command will be in /usr/local/bin
and should be usable. Try it. From the FirstBase MAIN menu, choose the
Application Menu by typing `app' and a <RETURN>.

You can play around with the two database areas, Rolodex and Phone,
and the sample source code area. Each has a menu.
The Rolodex application demonstrates most every aspect of getting
at data stored in a FirstBase database.

For documentation on FirstBase, start with the Postscript manual
located in $FIRSTBASEHOME/doc/fb-manual.ps.

Bugs and Comments
-----------------

Please report any bugs to fb-bug@firstbase.com.
Please send comments to fb-comments@firstbase.com.

---
$Id: README,v 9.4 2001/01/22 18:03:06 john Exp $


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/include/fb.h
/*
* Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
*
* $Id: fb.h,v 9.2 2006/04/03 17:28:59 john Exp $
*
* See the file LICENSE for conditions of use and distribution.
*
*/

#if HAVE_CONFIG_H
#include <config.h>
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <ctype.h>
#include <math.h>
#if HAVE_UNISTD_H
#include <sys/types.h>
#include <unistd.h>
#endif /* HAVE_UNISTD_H */
#include <stdlib.h>
#include <sys/stat.h>
#include <errno.h>

/* simple macros */
#define FB_EDITMODE	1
#define FB_ENDMODE	0
#define FB_WITHROOT	1
#define FB_NOROOT	0
#define FB_FDIGITS	(SYSMSG[S_FORMAT_DIGITS])
#define FB_FSTRING	(SYSMSG[S_FORMAT_STRING])
#define FB_FCHAR	(SYSMSG[S_FORMAT_CHAR])
#define FB_F_READ	(SYSMSG[S_F_READ])
#define FB_F_WRITE	(SYSMSG[S_F_WRITE])
#define FB_FDATE	(SYSMSG[S_FDATE])
#define FB_BLANK	CHAR_BLANK
#define FB_NEWLINE	CHAR_NEWLINE
#define FB_HELP 	CHAR_QUESTION
#define FB_INTO 	CHAR_G_THAN
#define FB_OUTOF	CHAR_L_THAN
#define FB_DEREF	CHAR_AT
#define FB_DOT		CHAR_DOT
#define FB_UNDERSCORE 	CHAR_UNDERSCORE

/* firstbase macros */
#define FB_CHECKERROR(x) if ((x) == FB_ERROR) return(FB_ERROR);
#define FB_FLD(f,db)	((db)->kp[(f)]->fld)
#define FB_TYP(t,db)	((db)->kp[(t)]->type)
#define FB_ISDELETED(db) (db->kp[db->nfields]->fld[0] == '*')
#define FB_RECMOD(db)	((db->kp[db->nfields]->fld) + 9)
#define FB_RECUMOD(db)	(db->kp[db->nfields]->fld[9])
#define FB_RECGMOD(db)	(db->kp[db->nfields]->fld[10])
#define FB_RECOMOD(db)	(db->kp[db->nfields]->fld[11])

#define FB_OFALPHA(t)	((t == FB_ALPHA) || (t == FB_STRICTALPHA) || \
			   (t == FB_UPPERCASE))
#define FB_OFNUMERIC(t)	((t == FB_DOLLARS) || (t == FB_FLOAT) || \
			   (t == FB_NUMERIC) || (t == FB_POS_NUM))
#define FB_SLONG	(sizeof(long))
#define FB_MAPREC_SIZE 	(sizeof(long) * 2)

/* numeric constants */
#define FB_CN_LEN 	25		/* company name length */
#define FB_HEADELEM 	10		/* single header element size */

#define FB_HEADSTART	4L		/* position of 1st headelem */
#define FB_LONGFIELD 	300		/* longest field editable by FB */
#define FB_MAXBY 	20		/* maximum sortby fields */
#define FB_MAXIDXS 	FB_MAXBY	/* maximum number fields/index */
#define FB_MAXKEEP 	1000		/* max num fields in pgen/lgen */
#define FB_MAXLINE 	250		/* general mac line length */
#define FB_MAXNAME 	240		/* maximum PATHNAME length */
#define FB_MODESTART	24L		/* start of uid gid mode for record */
#define FB_NUMLENGTH 	25		/* width of numeric typed fields */
#define FB_PRINTLINE	501
#define FB_PSIZE 	10		/* password size */
#define FB_RECORDPTR	10		/* length of record pointer */
#define FB_SCREENFIELD 	50		/* screen window: normal fields */
#define FB_SEQSIZE 	4		/* sequence id number size */
#define FB_SEQSTART	0L		/* position of sequence id */
#define FB_TITLESIZE 	10		/* max size of fld title (id) */
#define FB_WAIT 	1		/* wait condition for lock */
#define FB_NOWAIT 	0		/* nowait condition for lock */
#define FB_AMERICAN	1		/* a date type */
#define FB_EUROPEAN	2		/* a date type */
#define FB_FADD 	16		/* just different than !@<> etc */

typedef struct fb_s_link fb_link;	/* defined link */
typedef struct fb_s_field fb_field;	/* defined field */
typedef struct fb_s_database fb_database;/* defined database */
typedef struct fb_s_bidx fb_bidx;	/* btree+ index set */
typedef struct fb_s_bseq fb_bseq;	/* btree+ sequence set */
typedef struct fb_s_ixauto fb_autoindex;/* autoindex structure */

struct fb_s_bidx {			/* btree+ index set node */
   long bi_left;			/* left side info */
   char bi_ltype;
   long bi_middle;			/* middle side info */
   char bi_mtype;
   long bi_right;			/* right side info */
   char bi_rtype;
   char *bi_key1;			/* separation keys */
   char *bi_key2;
   int bi_lock;
   char *bi_name;			/* index name */
   int bi_fd;				/* index set file descriptor */
   char *bi_rec;			/* storage for index record */
   int bi_ksize;			/* size of an index key */
   long bi_root;			/* index root */
   long bi_height;			/* index height */
   long bi_recsiz;			/* index record size */
   long bi_recno;			/* current record number */
   long bi_reccnt;			/* index record count */
   long bi_free;			/* free pointer */
   };

struct fb_s_bseq {			/* btree+ sequence set node */
   long bs_prev;			/* prev and next pointers */
   long bs_next;
   char *bs_key1;			/* index keys */
   char *bs_key2;
   char *bs_key3;
   int bs_curkey;			/* current key pointer */
   int bs_lock;
   char *bs_name;			/* sequence file name */
   int bs_fd;				/* sequence file descriptor */
   char *bs_rec;			/* storage for sequnce record */
   int bs_ksize;			/* size of a sequence key */
   long bs_head;			/* head of sequence list */
   long bs_tail;			/* tail of sequence list */
   long bs_recsiz;			/* size of sequence record */
   long bs_recno;			/* current sequence record number */
   long bs_reccnt;			/* sequence record count */
   long bs_free;			/* free pointer */
   };

struct fb_s_link {
   fb_database *f_dp;			/* pointer to linked database */
   fb_field *f_fix;			/* indexed value in 'from' database */
   fb_field *f_tix;			/* ix val in 'to' db providing link */
   fb_field *f_ffp;			/* field within 'from' database */
   fb_field *f_tfp;			/* field within 'to' database */
   char *f_fld;				/* storage area for field val */
   char *f_xfld;			/* index value creating field link */
   fb_link *f_next;			/* to link all flinks for searching */
   long f_absrec;			/* absolute record number to use */
   fb_database *f_basedp;		/* database which referred link */
   };

struct fb_s_ixauto {			/* auto indexed field struct */
   char *autoname;			/* for storing autoindex name */
   char *dup_fld;			/* for 'old' field value */
   int afd, hfd;			/* auto index and header fds */
   short uniq;				/* uniqness flag */

   int ix_tree;				/* btree flag */
   fb_bseq *ix_seq, *ix_seqtmp;		/* btree sequence set */
   fb_bidx *ix_idx, *ix_idxtmp;		/* btree index set */
   fb_field **ix_ip;			/* field pointers array for auto ix */
   long ix_bsmax, ix_bsend;		/* for btrees - for auto ix array */
   int ix_ifields;			/* number of fields in the ix array */
   char *ix_key_fld;			/* key space for index entry */
   };

struct fb_s_field {				/* a database field */
   char *id;				/* unique title labeling field */
   char type;				/* type of the field */
   int size;				/* maximum size of field */
   int loc;				/* location when arec is used */
   char *fld;				/* actual pointer to field */
   int incr;				/* to store incremental defaults */
   char *comment, *idefault, *help, *prev, *range, *a_template, *f_macro;
   char comloc, lock, choiceflag;	/* various flags */
   fb_autoindex *aid;			/* pointer to auto index data */
   fb_link *dflink;			/* link to other database */
   fb_link *xflink;			/* extended choice/link database */
   char *mode;				/* permissions settings per field */
   short int f_prec;			/* precision for field */
   };

struct fb_s_database {			/* FB database header record */
   char *dbase;				/* full database path names */
   char *dindex;
   char *dmap;
   char *ddict;
   char *idict;
   char *sdict;				/* simple screen name */
   char *dlog;				/* log name of database */
   long reccnt;				/* full record count */
   char dirty;				/* dirty bit == 0 or 1 */
   long delcnt;				/* deleted record count */
   long rec;				/* current record number */
   int recsiz;				/* approximate record size */
   int nfields;				/* number of fields */
   int fd;				/* database file descriptor */
   int ifields;				/* number of index fields */
   int ifd;				/* index file descriptor */
   int ihfd;				/* index header file descriptor */
   int logfd;				/* log file descriptor */
   int irecsiz;				/* index record size */
   long bsmax, bsend, bsrec;		/* search max/end points */
   int mfd;				/* record map file descriptor */
   fb_field **kp;			/* array of fields */
   fb_field **ip;			/* array of index fields */
   char *orec;				/* original copy of rec */
   char *arec;				/* alternate copy of rec */
   char *irec;				/* buffer for an index record */
   char *afld;				/* field buffers for this db */
   char *bfld;				/* field buffers for this db */
   int refcnt;				/* reference count for linktop list */

					/* B+tree index fields */
   int b_tree;				/* flag set if dbase is using btrees */
   fb_bseq *b_seq, *b_seqtmp;		/* sequence set (and temp space) */
   fb_bidx *b_idx, *b_idxtmp;		/* index set (and temp space) */
   fb_autoindex **b_autoindex;		/* autoindex ptr array */
   int b_maxauto;			/* max number of auto index */
   int b_curauto;			/* cur number of auto index array */
   short int fixedwidth;		/* fixed width flag */

   int b_sid;				/* server id - open via server */
   short int inside_bulkrec;		/* server id - open via server */
   int sequence;			/* sequence number - ala fb_getseq() */
   };

/* external data - always included - defined in sysmsg.c */
extern char *SYSMSG[];

#include <input.h>
#include <error.h>
#include <message.h>
#include <mdict.h>

#include <dbpwd.h>
#include <vdict.h>

/* Cdb to X header file */
#include <cx.h>

#include <libcc.h>
#include <libcdb.h>
#include <libdbd.h>
#include <libinit.h>
#include <liblic.h>
#include <librec.h>
#include <libscr.h>
#include <libwin.h>
#include <libsec.h>
#include <libedit.h>
#include <libclnt.h>
#include <interface.h>
